web/sw.js

```diff
@@ -0,0 +1,95 @@
+// There can be multiple clients (pages) receiving files, so they generate an id
+// and here we store info assosiated with each transfer.
+const streams = new Map();
+
+// Repurposing the map...
+const waitForMetadata = id => new Promise((resolve, reject) => {
+	streams.set(id, {resolve, reject});
+});
+const signalMetadataReady = (id, streamInfo) => {
+	if (streams.has(id)) streams.get(id).resolve(streamInfo);
+}
+
+const createStream = (onCancel) => {
+	const streamInfo = {};
+
+	streamInfo.stream = new ReadableStream({
+		start (controller) {
+			streamInfo.controller = controller;
+		},
+		cancel (reason) {
+			onCancel(reason);
+		}
+	});
+
+	return streamInfo;
+};
+
+self.addEventListener('message', event => {
+	const message = event.data;
+	const id = message.id;
+
+	if (message.type === 'metadata') {
+		const {name, size} = message;
+
+		// TODO propagate cancellation back to main window and sender.
+		const onCancel = cancelReason => console.warn('Stream cancelled', cancelReason);
+		const streamInfo = { name, size, ...createStream(onCancel) };
+
+		// Resolve promise if GET request arrived first.
+		signalMetadataReady(id, streamInfo);
+
+		streams.set(id, streamInfo);
+
+	} else {
+		const streamInfo = streams.get(id);
+
+		if (message.type === 'data') {
+			streamInfo.controller.enqueue(new Uint8Array(message.data));
+
+		} else if (message.type === 'end') {
+			streamInfo.controller.close();
+
+			// Synchronize with fetch handler to clean up properly.
+			if (streamInfo.requestHandled)
+				streams.delete(id);
+			else streamInfo.streamHandled = true;
+
+		} else if (message.type === 'error') {
+			streamInfo.controller.error(message.error);
+		}
+	}
+});
+
+const encodeFilename = filename =>
+	encodeURIComponent(filename).replace(/['()]/g, escape).replace(/\*/g, '%2A');
+
+self.addEventListener('fetch', (event) => {
+	const url = new URL(event.request.url);
+
+	if (event.request.method !== 'GET' || !url.pathname.match(/^\/download\/[^\/]+$/)) return;
```

Comment:
can this ever happen?

Reply:
It's a kind of a sanity test right now. I couldn't get the ServiceWorker with `/donwload/` scope working for some reason and used global one for some time.